File: UnityScript/Example/Assets/OpenPath/OPNavMesh.js
#pragma strict

import System.Collections.Generic;

class OPNavMesh extends MonoBehaviour {	
	private class Triangle {
		var indices : int[];
		
		function Triangle ( v0 : int, v1 : int, v2 : int ) {
			indices = new int[3];
			indices[0] = v0;
			indices[1] = v1;
			indices[2] = v2;
		}
		
		private function IsNeighborTo ( t : Triangle, vertices : Vector3[] ) : boolean {
			var similarVertices : int = 0;
			
			for ( var thisVertex : int in indices ) {
				for ( var thatVertex : int in t.indices ) {
					if ( vertices[thisVertex] == vertices[thatVertex] ) {
						similarVertices++;
					}
				}
			}
			
			return similarVertices > 1;
		}
		
		public function GetNeighbors ( triangles : Triangle[], vertices : Vector3[] ) : List.< int > {
			var tempList : List.< int > = new List.< int > ();
			
			for ( var i : int = 0; i < triangles.Length; i++ ) {
				if ( IsNeighborTo ( triangles[i], vertices ) ) {
					tempList.Add ( i );
				}
			}
			
			return tempList;
		}
		
		public function GetMedianPoint ( mesh : Mesh ) : Vector3 {
			var result : Vector3;
			
			for ( var i : int = 0; i < indices.Length; i++ ) {
				result += mesh.vertices[indices[i]];
			}
			
			result /= indices.Length;
		
			return result;
		}
	}
	
	private function MakeNeighbors ( a : OPNode, b : OPNode ) {
		if ( !a.neighbors.Contains ( b ) ) {
			a.neighbors.Add ( b );
		}
		
		if ( !b.neighbors.Contains ( a ) ) {
			b.neighbors.Add ( a );
		}
	}
	
	public function GetNodes () : OPNode[] {
		var mesh : Mesh = this.GetComponent(MeshFilter).mesh;
		var triangleList : List.< Triangle > = new List.< Triangle > ();
		var allNodes : List.< OPNode > = new List.< OPNode > ();
		
		var i : int = 0;
		var nb : int = 0;

		// Create triangles
		for ( i = 0; i < mesh.triangles.Length; i += 3 ) {			
			var triangle : Triangle = new Triangle (
				mesh.triangles [ i ],
				mesh.triangles [ i + 1 ],
				mesh.triangles [ i + 2 ]
			);
						
			triangleList.Add ( triangle );
			
			// Create median node
			var mn : OPNode = new OPNode ();
			mn.position = this.transform.TransformPoint ( triangle.GetMedianPoint ( mesh ) );
			
			// Add median node to list
			allNodes.Add ( mn );
		}
		
		var triangleArray : Triangle[] = triangleList.ToArray();
		var vertices : Vector3[] = mesh.vertices;
		
		// Connect median nodes
		for ( i = 0; i < triangleArray.Length; i++ ) {		
			for ( nb in triangleArray[i].GetNeighbors ( triangleArray, vertices ) ) {
				MakeNeighbors ( allNodes [ i ], allNodes [ nb ] );
			}
		}
		
		// Return
		return allNodes.ToArray();
	}
}

File: C#/Example/Assets/OpenPath/OPNavMesh.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class OPNavMesh : MonoBehaviour {

    private class Triangle
    {
        int[] indicies;

        public Triangle(int v0, int v1, int v2)
        {
            indicies = new int[3];
            indicies[0] = v0;
            indicies[1] = v1;
            indicies[2] = v2;
        }

        private bool IsNeighborTo(Triangle t, Vector3[] vertices)
        {
            int similarVerticies = 0;

            foreach (int thisVertex in indicies)
            {
                foreach (int thatVertex in t.indicies)
                {
                    if (vertices[thisVertex] == vertices[thatVertex])
                    {
                        similarVerticies++;
                    }
                }
            }
            return similarVerticies > 1;
        }

        public List<int> GetNeighbors(Triangle[] triangles, Vector3[] vertices)
        {
            List<int> tempList = new List<int>();

            for (int i = 0; i < triangles.Length; i++)
            {
                if (IsNeighborTo(triangles[i], vertices))
                {
                    tempList.Add(i);
                }
            }
            return tempList;
        }

        public Vector3 GetMedianPoint(Mesh mesh)
        {
            Vector3 result = new Vector3();

            for (int i = 0; i < indicies.Length; i++)
            {
                result += mesh.vertices[indicies[i]];
            }
            result /= indicies.Length;
            return result;
        }
    }

    private void MakeNeighbors(OPNode a, OPNode b)
    {
        if (!a.neighbors.Contains(a))
        {
            b.neighbors.Add(b);
        }

        if (!b.neighbors.Contains(a))
        {
            b.neighbors.Add(a);
        }
    }

    public OPNode[] GetNodes()
    {
        Mesh mesh = this.GetComponent<MeshFilter>().mesh;
        List<Triangle> triangleList = new List<Triangle>();
        List<OPNode> allNodes = new List<OPNode>();

        int i = 0;

        //create triangles
        for (i = 0; i < mesh.triangles.Length; i += 3)
        {
            Triangle triangle = new Triangle(mesh.triangles[i], mesh.triangles[i+1], mesh.triangles[i+2]);

            triangleList.Add(triangle);

            //create median node
            OPNode mn = new OPNode();
            mn.position = this.transform.TransformPoint(triangle.GetMedianPoint(mesh));

            //add median node to list
            allNodes.Add(mn);
        }

        Triangle[] triangleArray = triangleList.ToArray();
        Vector3[] verticies = mesh.vertices;

        //connect median nodes
        for (i = 0; i < triangleArray.Length; i++)
        {
            foreach( int nb in triangleArray[i].GetNeighbors(triangleArray, verticies)){
                MakeNeighbors(allNodes[i], allNodes[nb]);
            }
        }
        //return
        return allNodes.ToArray();
    }
}
